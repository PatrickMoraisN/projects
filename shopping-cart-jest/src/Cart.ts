import { remove, find } from 'lodash';

type ItemProps = {
  quantity: number;
  product: ProductProps;
}

type ProductProps = {
  title: string;
  price: number;
}

export default class Cart {
  items = [];

  getTotal() {
    return this.items.reduce((acc, item) => {
      return acc + item.quantity * item.product.price;
    }, 0);
  }

  add(item: ItemProps) {
    const itemToFind = { product: item.product };

    if (find(this.items, itemToFind)) {
      remove(this.items, itemToFind);
    }
    this.items.push(item);
  }

  remove(product: ProductProps) {
    const itemToFind = { product };
    remove(this.items, itemToFind);
  }

  summary() {
    const total = this.getTotal();
    const items = this.items;
  
    return {
      total,
      items,
    }
  }

  checkout() {
    const { total, items } = this.summary();

    this.items = [];
  
    return {
      total,
      items,
    }
  }
}

const cart = new Cart();

const product = {
  title: '',
  price: 0,
};

const item = {
  quantity: 2,
  product,
};

cart.add(item);
cart.remove(product);
cart.summary();
cart.checkout();