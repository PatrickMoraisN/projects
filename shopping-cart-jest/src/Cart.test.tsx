import Cart from './Cart';

describe('Cart', () => {
  let cart;
  beforeEach(() => {
    cart = new Cart();
  })

  it('should return 0 when getTotal() is executed in a newly created instance', () => {

    expect(cart.getTotal()).toBe(0);
  });

  it('should multiply quantity and price and receive the total amount', () => {
    const item = {
      product: {
        title: 'Nike run - men',
        price: 35388,
      },
      quantity: 3,
    };

    cart.add(item);

    expect(cart.getTotal()).toBe(106164);
  });
})
