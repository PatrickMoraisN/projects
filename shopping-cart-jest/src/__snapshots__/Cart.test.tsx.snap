// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Nike run - men",
      },
      "quantity": 3,
    },
    Object {
      "product": Object {
        "price": 65388,
        "title": "Nike run - women",
      },
      "quantity": 1,
    },
  ],
  "total": 171552,
}
`;

exports[`Cart checkout() should return an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Nike run - men",
      },
      "quantity": 3,
    },
    Object {
      "product": Object {
        "price": 65388,
        "title": "Nike run - women",
      },
      "quantity": 1,
    },
  ],
  "total": 171552,
}
`;

exports[`Cart checkout() should returnan object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Nike run - men",
      },
      "quantity": 3,
    },
    Object {
      "product": Object {
        "price": 65388,
        "title": "Nike run - women",
      },
      "quantity": 1,
    },
  ],
  "total": 171552,
}
`;
